File: packages/blendlink/src/knownIssues.ts
import { existsSync, readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'
import {
  compareBlenderVersions,
  parseBlenderVersion,
  type BlenderVersion,
} from './blenderCompatibility.js'

export type KnownIssueSeverity = 'warning' | 'block'

export interface BlenderKnownIssue {
  id: string
  summary: string
  action: string
  severity: KnownIssueSeverity
  minInclusive: string
  maxExclusive: string
  evidence: string[]
}

export interface BlenderKnownIssueRegistry {
  schemaVersion: 1
  policy: string
  issues: BlenderKnownIssue[]
}

const BLENDER_EVIDENCE_HOSTS = new Set([
  'docs.blender.org',
  'developer.blender.org',
  'projects.blender.org',
])

function isPrimaryEvidence(url: URL): boolean {
  return BLENDER_EVIDENCE_HOSTS.has(url.hostname) || (
    url.hostname === 'github.com' &&
    url.pathname.startsWith('/KhronosGroup/glTF-Blender-IO/')
  )
}

function text(value: unknown, field: string): string {
  if (typeof value !== 'string' || !value.trim()) throw new Error(`${field} must be non-empty text`)
  return value.trim()
}

/** Validate the checked-in registry as product data, not executable code. */
export function parseBlenderKnownIssueRegistry(input: unknown): BlenderKnownIssueRegistry {
  if (!input || typeof input !== 'object' || Array.isArray(input)) {
    throw new Error('known-issue registry must be an object')
  }
  const root = input as Record<string, unknown>
  if (root.schemaVersion !== 1) {
    throw new Error(`known-issue registry schemaVersion ${String(root.schemaVersion)} is unsupported (expected 1)`)
  }
  if (!Array.isArray(root.issues)) throw new Error('known-issue registry issues must be an array')
  const ids = new Set<string>()
  const issues = root.issues.map((candidate, index) => {
    if (!candidate || typeof candidate !== 'object' || Array.isArray(candidate)) {
      throw new Error(`issues[${index}] must be an object`)
    }
    const item = candidate as Record<string, unknown>
    const id = text(item.id, `issues[${index}].id`)
    if (!/^[a-z0-9][a-z0-9-]*$/.test(id)) throw new Error(`issues[${index}].id must be a lowercase slug`)
    if (ids.has(id)) throw new Error(`duplicate known-issue id ${id}`)
    ids.add(id)
    const severity = item.severity
    if (severity !== 'warning' && severity !== 'block') {
      throw new Error(`issues[${index}].severity must be warning or block`)
    }
    const minInclusive = text(item.minInclusive, `issues[${index}].minInclusive`)
    const minimum = parseBlenderVersion(minInclusive, `issues[${index}].minInclusive`)
    const maxExclusive = text(item.maxExclusive, `issues[${index}].maxExclusive`)
    if (compareBlenderVersions(
      parseBlenderVersion(maxExclusive, `issues[${index}].maxExclusive`),
      minimum,
    ) <= 0) {
      throw new Error(`issues[${index}].maxExclusive must be newer than minInclusive`)
    }
    if (!Array.isArray(item.evidence) || item.evidence.length === 0) {
      throw new Error(`issues[${index}].evidence needs at least one primary-source URL`)
    }
    const evidence = item.evidence.map((source, sourceIndex) => {
      const value = text(source, `issues[${index}].evidence[${sourceIndex}]`)
      let url: URL
      try {
        url = new URL(value)
      } catch {
        throw new Error(`issues[${index}].evidence[${sourceIndex}] is not a URL`)
      }
      if (url.protocol !== 'https:' || !isPrimaryEvidence(url)) {
        throw new Error(
          `issues[${index}].evidence[${sourceIndex}] must be an HTTPS Blender/Khronos primary source`,
        )
      }
      return value
    })
    return {
      id,
      summary: text(item.summary, `issues[${index}].summary`),
      action: text(item.action, `issues[${index}].action`),
      severity,
      minInclusive,
      maxExclusive,
      evidence,
    } satisfies BlenderKnownIssue
  })
  return {
    schemaVersion: 1,
    policy: text(root.policy, 'known-issue registry policy'),
    issues,
  }
}

function registryPath(): string {
  const here = dirname(fileURLToPath(import.meta.url))
  const candidates = [
    join(here, 'addon', 'blender_known_issues.json'),
    join(here, '..', '..', 'blender-addon', 'blender_known_issues.json'),
  ]
  const path = candidates.find(existsSync)
  if (!path) {
    throw new Error(
      `bundled Blender known-issue registry is missing (checked ${candidates.join(', ')})`,
    )
  }
  return path
}

export function loadBlenderKnownIssueRegistry(): BlenderKnownIssueRegistry {
  const path = registryPath()
  try {
    return parseBlenderKnownIssueRegistry(JSON.parse(readFileSync(path, 'utf8')))
  } catch (error) {
    throw new Error(
      `${path}: ${error instanceof Error ? error.message : String(error)}`,
    )
  }
}

export function matchingBlenderKnownIssues(
  blenderVersion: BlenderVersion,
  registry = loadBlenderKnownIssueRegistry(),
): BlenderKnownIssue[] {
  return registry.issues.filter((issue) => {
    if (compareBlenderVersions(
      blenderVersion,
      parseBlenderVersion(issue.minInclusive, `${issue.id}.minInclusive`),
    ) < 0) return false
    return compareBlenderVersions(
      blenderVersion,
      parseBlenderVersion(issue.maxExclusive, `${issue.id}.maxExclusive`),
    ) < 0
  })
}

File: packages/blender-addon/known_issues.py
# SPDX-License-Identifier: GPL-3.0-or-later
"""Evidence-gated Blender-version warnings shared by the addon UI.

The JSON beside this module is intentionally data, not Python behavior. Empty
is a valid state: Blendlink only warns when a bounded failure has public primary
evidence and a practical artist action.
"""
from __future__ import annotations

import json
import re
from pathlib import Path
from urllib.parse import urlparse

REGISTRY_PATH = Path(__file__).with_name("blender_known_issues.json")
BLENDER_EVIDENCE_HOSTS = {
    "docs.blender.org",
    "developer.blender.org",
    "projects.blender.org",
}
_registry_cache = None
_registry_error = None


def _version(value, field):
    if not isinstance(value, str):
        raise ValueError(f"{field} must be a numeric Blender version")
    parts = value.split(".")
    if len(parts) not in (2, 3) or any(not part.isdigit() for part in parts):
        raise ValueError(f"{field} must be a numeric Blender version such as 4.2 or 5.1.1")
    values = tuple(int(part) for part in parts)
    return values + (0,) * (3 - len(values))


def _text(value, field):
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{field} must be non-empty text")
    return value.strip()


def validate_registry(data):
    if not isinstance(data, dict) or data.get("schemaVersion") != 1:
        raise ValueError("known-issue registry schemaVersion must be 1")
    _text(data.get("policy"), "known-issue registry policy")
    issues = data.get("issues")
    if not isinstance(issues, list):
        raise ValueError("known-issue registry issues must be an array")
    seen = set()
    for index, issue in enumerate(issues):
        if not isinstance(issue, dict):
            raise ValueError(f"issues[{index}] must be an object")
        issue_id = _text(issue.get("id"), f"issues[{index}].id")
        if re.fullmatch(r"[a-z0-9][a-z0-9-]*", issue_id) is None:
            raise ValueError(f"issues[{index}].id must be a lowercase slug")
        if issue_id in seen:
            raise ValueError(f"duplicate known-issue id {issue_id}")
        seen.add(issue_id)
        _text(issue.get("summary"), f"issues[{index}].summary")
        _text(issue.get("action"), f"issues[{index}].action")
        if issue.get("severity") not in {"warning", "block"}:
            raise ValueError(f"issues[{index}].severity must be warning or block")
        minimum = _version(issue.get("minInclusive"), f"issues[{index}].minInclusive")
        maximum = _text(issue.get("maxExclusive"), f"issues[{index}].maxExclusive")
        if _version(maximum, f"issues[{index}].maxExclusive") <= minimum:
            raise ValueError(f"issues[{index}].maxExclusive must be newer than minInclusive")
        evidence = issue.get("evidence")
        if not isinstance(evidence, list) or not evidence:
            raise ValueError(f"issues[{index}].evidence needs at least one primary-source URL")
        for source_index, source in enumerate(evidence):
            parsed = urlparse(_text(source, f"issues[{index}].evidence[{source_index}]"))
            primary = parsed.hostname in BLENDER_EVIDENCE_HOSTS or (
                parsed.hostname == "github.com"
                and parsed.path.startswith("/KhronosGroup/glTF-Blender-IO/")
            )
            if parsed.scheme != "https" or not primary:
                raise ValueError(
                    f"issues[{index}].evidence[{source_index}] must be an HTTPS "
                    "Blender/Khronos primary source"
                )
    return data


def load_registry(path=REGISTRY_PATH):
    try:
        return validate_registry(json.loads(Path(path).read_text(encoding="utf8")))
    except (OSError, json.JSONDecodeError, ValueError) as error:
        raise RuntimeError(f"{path}: {error}") from error


def prime_cache() -> None:
    """Read and validate the bundled registry outside panel draw calls."""
    global _registry_cache, _registry_error
    try:
        _registry_cache = load_registry()
        _registry_error = None
    except RuntimeError as error:
        _registry_cache = None
        _registry_error = error


def matching(version, registry=None):
    version = tuple(version[:3])
    data = registry if registry is not None else load_registry()
    matches = []
    for issue in data["issues"]:
        if version < _version(issue["minInclusive"], f"{issue['id']}.minInclusive"):
            continue
        if version >= _version(issue["maxExclusive"], f"{issue['id']}.maxExclusive"):
            continue
        matches.append(issue)
    return matches


def current():
    import bpy
    if _registry_cache is None and _registry_error is None:
        # Direct module use in tests or source sessions that bypass register().
        prime_cache()
    if _registry_error is not None:
        raise RuntimeError(str(_registry_error))
    return matching(bpy.app.version, _registry_cache)
